### PR TITLE
MarianMTModel.from_pretrained('Helsinki-NLP/opus-marian-en-de')

### LINKED ISSUE #3775: OpusNMT/MarianMT Machine Translation Models
### Model description
1,026 Language Pair Models, downloadable [here](http://opus.nlpl.eu/Opus-MT/)
Trained with Marian C++ [library](https://github.com/marian-nmt/marian)

### Open source status

* [ x] the model implementation is available: (give details)
* [ x] the model weights are available: (give details)
* [ x] who are the authors: TODO, find gh-usernames of authors!

### Proposed API:
```python
model_name = 'marian/en-fr'
tokenizer = MarianTokenizer.from_pretrained(model_name)
model = MarianModel.from_pretrained(model_name)
src_text = "how are you today?"
tgt_text = "comment allez vous aujourd'hui?"
# Training API
full_inputs: dict = tokenizer.prepare_batch_for_translation(src_text, tgt_text=tgt_text)
loss, logits, *other_outputs = model(
        full_inputs['input_ids'],
        full_inputs['attention_mask'],

        full_inputs['decoder_input_ids'], # this argument is mandatory for the forward pass
)
# Inference/generate API
src_inputs: dict = tokenizer.prepare_batch_for_translation(src_text)
generated_fr_ids = model.generate(src_inputs['input_ids'], src_inputs['attention_mask'])
french_text: List[str] = processor.decode_batch(generated_fr_ids)
```

### Implementation Details
`MarianTokenizer` Signatures

(Not calling it Tokenizer to avoid confusion, but I don't feel strongly.)
- All models require `MosesSentenceSplitter` and `MosesPunctuationNormalizer` preprocessing
    -  There are some additional perl scripts we will not port for pre/post-processing
- 81 of the models require BPE, 960 require SentencePiece.
- We can decide which is which

```python
class MarianTokenizer:
    def __init__(self, vocab_file, source_bpe, target_bpe, source_spm, target_spm):
        # decide whether to use BPE/SPM based on which files are present in S3.
        self.source_lang, self.target_lang #inferred from paths/config
        #self.source_spm =

    @property
    def uses_sentencepiece(self) -> bool:
        return self.source_spm is not None

    def from_pretrained(self, *args, **kwargs):
        # needs to be overwritten or modified to not fail if certain files not present.

    def prepare_batch_for_translation(src_text:str, tgt_text=None, return_tensors='pt', max_length=512, pad_to_max_length=True) -> Dict[str, tensor/list]:
		return {}


    def decode_batch(self, target_lang_ids: List[List[int]]) -> List[str]:


    def decode(self, target_lang_id: List[int]) -> str:
```

#### Edits
- renamed `MarianProcessor` -> `MarianTokenizer` for consistency

### PR DESCRIPTION
**Pull request by sshleifer**
Adds support for `opus/marian-en-de` translation models:
- There are 900 models with this `MarianSentencePieceTokenizer`, `MarianMTModel` setup. - This PR only adds en-de to avoid massive S3 maintenance if names/other things change.
- There is no formal connection to the bart authors, but the bart code is well-tested and fast and I didn't want to rewrite it. You can still read everything in one file: modeling_bart.py.
- unittest failures are from isort.

The only differences from BART are:
- static (sinusoid) positional embeddings (`config.static_position_embeddings`)
- a new `final_logits_bias` (`config.add_bias_logits`)
- no `layernorm_embedding`  (`config.normalize_embedding`)

I could have also used T5 but that has more prefix token logic and relative position bias.
Solutions that split the changes across two files were much harder to read.

Why MarianSentencePieceTokenizer instead of MarianTokenizer?
- about 10% of the models (another 100) use BPE instead of SentencePiece. We will use that tokenizer (or integrate it with this one and rename) in a future PR.

### Future PRs
- load in many models
- support BPE tokenizer in some way.
- make lm_labels on the fly for training loop, like T5.
- `TranslationPipeline`

`TODO[Code]`:

- [x] investigate discrepancies in generation parameters: `length_penalty`, `decoder_start_token_id`, etc. These may explain small differences in generations (detailed in comment below). 
- [x] isort

`TODO[Docs]`:

- [x] Naming: `MarianMTModel`, `MarianSentencePieceTokenizer`
- [x] put en-de model in S3, update integration tests.
- [ ] docstring examples
- [ ] notebook to show?
- [ ] marian.rst
- [ ] README.md
- [ ] `AutoModelWithLMHead`

### CODE REVIEW

File: tests/test_modeling_bart.py [hunk 1]
```diff
@@ -109,7 +110,9 @@ class BARTModelTest(ModelTesterMixin, unittest.TestCase):
     all_model_classes = (
         (BartModel, BartForConditionalGeneration, BartForSequenceClassification) if is_torch_available() else ()
     )
-    all_generative_model_classes = (BartForConditionalGeneration,) if is_torch_available() else ()
+    all_generative_model_classes = (
+        (BartForConditionalGeneration, MarianForConditionalGeneration) if is_torch_available() else ()
```
Comment (sshleifer) [inline tests/test_modeling_bart.py:114]: very cheap test coverage: + 2 seconds. Happy to delete.

File: src/transformers/sinusoidal_positional_embeddings.py [hunk 2]
```diff
@@ -0,0 +1,38 @@
+import numpy as np
+import torch
+from torch import nn
+
+
+class SinusoidalPositionalEmbedding(nn.Embedding):
```
Comment (patrickvonplaten) [inline src/transformers/sinusoidal_positional_embeddings.py:6]: Do we need an extra file for this or could we leave it in Bart for the moment?

**Comment by patrickvonplaten** [comment]
I think that's awesome! Such little code changes to include that many models is amazing :-) 
Does the model with the pre-trained weights produce 1-to-1 the same results as the official Marian models in C++?

File: src/transformers/file_utils.py [hunk 3]
```diff
@@ -548,3 +548,18 @@ def wrapper(*args, **kwargs):
             raise ImportError(f"Method `{func.__name__}` requires TF.")
 
     return wrapper
+
+
+def unzip(zip_path, dest_dir):
```
Comment (julien-c) [inline src/transformers/file_utils.py:553]: add type hints to functions in file_utils

**Comment by sshleifer** [comment]
@patrickvonplaten the generations are very sensible, but not identical to marian, due to slightly different generation parameters: 

Equivalent (4/7):

C+: Ich bin ein kleiner Frosch.
HF: Ich bin ein kleiner Frosch.

C+: Tom bat seinen Lehrer um Rat.
HF: Tom bat seinen Lehrer um Rat.

C+: Tom bewunderte Marias Mut wirklich.
HF: Tom bewunderte Marias Mut wirklich.

C+: Umdrehen und die Augen schließen.
HF: Umdrehen und die Augen schließen.

Not Equivalent (3/7):
C+: Jetzt kann ich die 100 Wörter vergessen, die ich kenne.
HF: Jetzt kann ich die 100 Wörter **des Deutschen** vergessen, die ich kenne.

C+: **O**  (Input="O")
HF: 

C+: So würde ich das **machen**.
HF: So würde ich das **tun**.

I'm investigating.

**Comment by patrickvonplaten** [comment]
> @patrickvonplaten the generations are very sensible, but not identical to marian, due to slightly different generation parameters:
> 
> Equivalent (4/7):
> 
> C+: Ich bin ein kleiner Frosch.
> HF: Ich bin ein kleiner Frosch.
> 
> C+: Tom bat seinen Lehrer um Rat.
> HF: Tom bat seinen Lehrer um Rat.
> 
> C+: Tom bewunderte Marias Mut wirklich.
> HF: Tom bewunderte Marias Mut wirklich.
> 
> C+: Umdrehen und die Augen schließen.
> HF: Umdrehen und die Augen schließen.
> 
> Not Equivalent (3/7):
> C+: Jetzt kann ich die 100 Wörter vergessen, die ich kenne.
> HF: Jetzt kann ich die 100 Wörter **des Deutschen** vergessen, die ich kenne.
> 
> C+: **O** (Input="O")
> HF:
> 
> C+: So würde ich das **machen**.
> HF: So würde ich das **tun**.
> 
> I'm investigating.

Sounds all very coherent :-)

File: src/transformers/sinusoidal_positional_embeddings.py (hunk 2 above)
Comment (patrickvonplaten) [reply src/transformers/sinusoidal_positional_embeddings.py:6]: I think there are multiple models which use SinusoidalPositionalEmbedding (distilbert, xlnet, ...) and they are all in their respective files. Just for consistency and it seems like too small to have its own file for me.

File: src/transformers/file_utils.py (hunk 3 above)
Comment (sshleifer) [reply src/transformers/file_utils.py]: Done!

File: src/transformers/sinusoidal_positional_embeddings.py (hunk 2 above)
Comment (sshleifer) [reply src/transformers/sinusoidal_positional_embeddings.py]: Done, makes sense.

File: src/transformers/modeling_bart.py [hunk 4]
```diff
@@ -859,6 +884,22 @@ def __init__(self, config: BartConfig):
...
+        self._resize_final_logits_bias(new_num_tokens, old_num_tokens)
+        return new_embeddings
+
+    def _resize_final_logits_bias(self, new_num_tokens: int, old_num_tokens: int) -> None:
+        if new_num_tokens <= old_num_tokens:
+            new_bias = self.final_logits_bias[:, :new_num_tokens]
+        else:
+            extra_bias = torch.zeros((1, new_num_tokens - old_num_tokens))
+            new_bias = torch.cat([self.final_logits_bias, extra_bias], dim=1)
+        self.register_buffer("final_logits_bias", new_bias)
```
Comment (LysandreJik) [inline src/transformers/modeling_bart.py:902]: Clean way to do this given that the bias and linear layer are not linked

File: src/transformers/modeling_marian.py [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+class MarianForConditionalGeneration(BartForConditionalGeneration):
+    """Pytorch version of marian-nmt's transformer.h (c++). Designed for the OPUS-NMT translation checkpoints.
+    Model API is identical to BartForConditionalGeneration"""
+
+    pretrained_model_archive_map = PRETRAINED_MODEL_ARCHIVE_MAP
+
+    def prepare_scores_for_generation(self, scores, cur_len, max_length):
+        if cur_len == max_length - 1 and self.config.eos_token_id is not None:
+            self._force_token_ids_generation(scores, self.config.eos_token_id)
+        return scores
```
Comment (LysandreJik) [inline src/transformers/modeling_marian.py:35]: Nice that it's this simple to have Marian after the changes you did to BART.

File: src/transformers/tokenization_marian.py [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+    "tokenizer_config_file": "tokenizer_config.json",
+}
+MODEL_NAMES = ("marian-en-de",)
+PRETRAINED_VOCAB_FILES_MAP = {
+    k: {m: f"{S3_BUCKET_PREFIX}/opus/{m}/{fname}" for m in MODEL_NAMES} for k, fname in vocab_files_names.items()
+}
+# Example URL https://s3.amazonaws.com/models.huggingface.co/bert/opus/marian-en-de/vocab.json
+
+
+class MarianSentencePieceTokenizer(PreTrainedTokenizer):
```
Comment (LysandreJik) [inline src/transformers/tokenization_marian.py:23]: Why `MarianSentencePieceTokenizer` instead of `MarianTokenizer`? Are we expecting other Marian tokenizers? (We don't have `AlbertSentencePieceTokenizer` or `XLNetSentencePieceTokenizer` even though they both are SP tokenizers)

File: src/transformers/tokenization_marian.py [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+        ids: list = self.convert_tokens_to_ids(tokens)[:max_length]
+        return ids + [self.eos_token_id]
+
+    def build_inputs_with_special_tokens(self, token_ids_0, token_ids_1=None) -> List[int]:
+        """
+        Build model inputs from a sequence or a pair of sequence for sequence classification tasks
+        by concatenating and adding special tokens.
+        A RoBERTa sequence has the following format:
+            single sequence: <s> X </s>
+            pair of sequences: <s> A </s></s> B </s>
```
Comment (LysandreJik) [inline src/transformers/tokenization_marian.py:104]: Doc seems wrong here (RoBERTa -> Marian?, sentence building doesn't seem to reflect the logic)

File: tests/test_modeling_bart.py (hunk 1 above)
Comment (LysandreJik) [reply tests/test_modeling_bart.py:115]: I think it's good to have this.

**Review by LysandreJik — COMMENTED** [review]
This looks nice! The conversion looks really complex, good job on that clean conversion script. Cool that there are integration tests.

File: src/transformers/modeling_bart.py [hunk 8]
```diff
@@ -227,6 +230,17 @@ def forward(self, x, encoder_padding_mask):
...
+    sq = x.squeeze()
+    if sq.ndim == 2:
+        slice = x[:3, :4]
+    elif sq.ndim == 3:
+        slice = x[:, 0, :6]
+    else:
+        slice = x[:5]
+    print(f"{msg}: shape: {x.shape} min: {x.min(): .4f} max: {x.max(): .4f} slice: {slice}")
+
+
```
[the remark this answers is not in the document]
Comment (sshleifer) [reply src/transformers/modeling_bart.py:243]: yes!

File: src/transformers/modeling_bart.py (hunk 4 above)
Comment (sshleifer) [reply src/transformers/modeling_bart.py:902]: Correct, final_bias_logits get overridden when weights are loaded. (Github doesn't want me reply to your upper comment)

File: src/transformers/tokenization_marian.py (hunk 6 above)
Comment (sshleifer) [reply src/transformers/tokenization_marian.py:23]: Good Q. About 10% of the Opus models don't use sentencepiece, they use BPE. Was going to tackle those in a separate PR.

File: src/transformers/tokenization_marian.py (hunk 7 above)
Comment (sshleifer) [reply src/transformers/tokenization_marian.py:104]: good catch!

File: src/transformers/convert_marian_to_pytorch.py [hunk 9]
```diff
@@ -92,6 +93,21 @@ def parse_readmes(repo_path="~/OPUS-MT-train/models"):
     return results
 
 
+def download_all_sentencepiece_models(repo_path="Opus-MT-train/models"):
+    """Requires 300GB"""
+    save_dir = Path("marian_ckpt")
+    if not Path(repo_path).exists():
+        raise ValueError("You must run: git clone git clone git@github.com:Helsinki-NLP/Opus-MT-train.git")
```
Comment (stefan-it) [inline src/transformers/convert_marian_to_pytorch.py:100]:
```suggestion
        raise ValueError("You must run: git clone git@github.com:Helsinki-NLP/Opus-MT-train.git")
```

File: src/transformers/file_utils.py [hunk 10]
```diff
@@ -548,3 +548,18 @@ def wrapper(*args, **kwargs):
...
+
+
+def save_json(content: Union[Dict, List], path: str) -> None:
+    with open(path, "w") as f:
+        json.dump(content, f)
+
+
+def load_json(path: str) -> Union[Dict, List]:
+    with open(path, "r") as f:
+        return json.load(f)
```
Comment (thomwolf) [inline src/transformers/file_utils.py:565]: Let's keep these in the conversion script?
I don't think we need to abstract 2 liners you only use in your conversion scripts.

File: src/transformers/tokenization_utils.py [hunk 11]
```diff
@@ -458,6 +458,14 @@ def char_to_word(self, batch_or_char_index: int, char_index: Optional[int] = Non
             char_index = batch_or_char_index
         return self._encodings[batch_index].char_to_word(char_index)
 
+    @torch_required
+    def to(self, device: str):
+        """Send all values to device by calling v.to(device)"""
+        assert is_torch_available(), "This function requires pytorch"
```
Comment (thomwolf) [inline src/transformers/tokenization_utils.py:464]: Do we need both the decorator and the assert?

File: tests/test_modeling_bart.py [hunk 12]
```diff
@@ -36,12 +36,14 @@
         BartConfig,
         BartTokenizer,
         MBartTokenizer,
+        MarianMTModel,
```
Comment (thomwolf) [inline tests/test_modeling_bart.py:39]: So, some of the tests of the `MarianMTModel` are in `test_modeling_bart` and some are in `test_modeling_marian`?
Looks a bit hard to remember to me.

**Review by thomwolf — APPROVED** [review]
Ok, looks great!

Awesome work @sshleifer!

Don't forget to update the model checkpoint naming as discussed before starting mass conversion.

File: tests/test_modeling_bart.py (hunk 12 above)
Comment (sshleifer) [reply tests/test_modeling_bart.py:39]: Yeah. I'll remove.

**Comment by sshleifer** [comment]
Merging. Will add docs in next PR, when I add tons of models.

**Comment by tholiao** [comment]
@sshleifer, can you elaborate on "investigate discrepancies in generation parameters: length_penalty, decoder_start_token_id, etc. These may explain small differences in generations (detailed in comment below)"? I see https://github.com/huggingface/transformers/pull/3908#issuecomment-618473222 but was this resolved?

**Comment by sshleifer** [comment]
_TLDR Was not resolved, the investigation continues._
Our lib has many parameters to control generations, `length_penalty`, `repetition_penalty`, `bad_word_ids`, `decoder_start_token_id`.
They are documented [here](https://huggingface.co/transformers/model_doc/bart.html?#transformers.BartForConditionalGeneration.generate).
The marian lib has `normalize` and `word-penalty`, which are not identical, but similar. Part of the discrepancy may involve configuring our parameters to be closer to theirs.

**Comment by tholiao** [comment]
I see, thanks!

**Comment by erikchwang** [comment]
Could anyone help with this issue: #5040 ?
